### PR TITLE
Tic Tac Toe | Rspec Testing

### PR DESCRIPTION
**Pull request by uzairali19**
# Tic Tac Toe Logic | Rspec Testing

Tic-tac-toe, noughts, and crosses is a paper-and-pencil game for two players, X and O, who take turns marking the spaces in a 3×3 grid. The player who succeeds in placing three of their marks in a diagonal, horizontal, or vertical row is the winner. It is a solved game with a forced draw assuming the best play from both players.

This branch is used to make a testing environment for Rspec to make sure all the game checks are complete.

## Strategy

A player can play a perfect tic-tac-toe game (to win or at least draw) if, each time it is their turn to play, they choose the first available move from the following list, as used in Newell Simon's 1972 tic-tac-toe program.[16]

- Win: If the player has two in a row, they can place a third to get three in a row.
- Block: If the opponent has two in a row, the player must play the third themselves to block the opponent.
- Fork: Create an opportunity where the player has two ways to win (two non-blocked lines of 2).
- Blocking an opponent's fork: If there is only one possible fork for the opponent, the player should block it. Otherwise, the player should block all forks in any way that simultaneously allows them to create two in a row. Otherwise, the player should create a two in a row to force the opponent into defending, as long as it doesn't result in them creating a fork. For example, if "X" has two opposite corners and "O" has the center, "O" must not play a corner move to win. (Playing a corner move in this scenario creates a fork for "X" to win.)
- Center: A player marks the center. (If it is the first move of the game, playing a corner move gives the second player more opportunities to make a mistake and may therefore be the better choice; however, it makes no difference between perfect players.)
- Opposite corner: If the opponent is in the corner, the player plays the opposite corner.
- Empty corner: The player plays in a corner square.
- Empty side: The player plays in a middle square on any of the 4 sides.

## Authors

👤 **Uzair**

- LinkedIn: [LinkedIn](https://www.linkedin.com/in/uzair-ali-964187166/)
- GitHub: [@githubhandle](https://github.com/uzairali19)

### CODE REVIEW

File: spec/tic_tac_toe_spec.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
+require 'board'
+
+# rubocop:disable Lint/UselessAssignment
+describe 'Game' do
+  g = GameBoard.new
```
Comment (bini-i) [inline spec/tic_tac_toe_spec.rb]: -Kindly use Rspec semantic `subject` to indicate that `g` is the object being tested

**Review by bini-i — CHANGES_REQUESTED** [review]
# Status: Changes Required :yellow_circle:            
### Hello @uzairali19  :wave: 

Very nice progress so far, 
- you have done a great job testing public methods available in your codebase.
- you have correctly used git-flow.
- you have placed your specs files in the correct directory.

#### To improve
- Kindly include both positive and negative scenarios for the tests you made.
> eg. Positive scenario:  it "gets user input and checks if input is converted to Integer"
          Negative scenario:  it "doesn't get user input and check if input is not converted to Integer"
- kindly use RSpec semantics like `subject` and `let`, 
> See inline comment below :point_down: 

#### If you have any questions feel free to ask within the PR and ping me on slack @Binyam Hailemeskel

### HAPPY CODING :computer: :blush: 

> #### Reviewed by Binyam Hailemeskel (Code Reviewer)
> GitHub: [@bini-i](https://github.com/bini-i) 
> Twitter: [@binyamshewa](https://twitter.com/binyamshewa)
> LinkedIn: [LinkedIn](https://www.linkedin.com/in/binyam-hailemeskel-728048151/)

**Review by carlos-ssh — CHANGES_REQUESTED** [review]
Hi @uzairali19, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Required Changes ♻️
Your project needs to cover positive and negative scenarios ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/Ruby_testing/commonsense.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59566098/))
Example: there is a spec for correct usage of the method [e.g. `` it "creates object if all required attributes are passed" ``]
and also there is spec for the wrong usage of a method [e.g.``  it "does not create the object if required attributes are missing" ``].

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have only 3 code reviews per project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by uzairali19** [comment]
> Hi @uzairali19,
> 
> Great work on making the changes requested by a previous reviewer 👏
> You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.
> 
> ## Required Changes ♻️
> _Check the comments under the review._
> 
> ## Optional suggestions
> _Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._
> 
> Cheers and Happy coding!👏👏👏
> 
> Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
> 
> **_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**
> 
> _As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have only 3 code reviews per project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

### Hello Carlos

Thank you for such a prompt response. However, the previous reviewer asked me to use `subject` and `let`. Just to address that concern I have used them in my code. Removing the outdated method of assigning `g` with `=`. I am attaching a screenshot for clarification. Please have a look at it. I would like my project to be approved.

Thank you 

Screenshot 2021-07-19 at 6 52 56 PM

Screenshot 2021-07-19 at 6 52 30 PM

File: spec/player_spec.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  let(:player) { Player.new('Uzair', 'Saad') }
+  context 'It Initializes the players' do
+    it 'Checks if the passed values are correct' do
+      player1 = player.player1
+      player2 = player.player2
+      expect(player1).to eq('Uzair')
+      expect(player2).to eq('Saad')
+    end
+  end
+end
```
Comment (george-shammar) [inline spec/player_spec.rb:13]: - [ ] This is a typical example of a negative test case for this test.
`it 'Checks if the passed values are incorrect' do
      player1 = player.player1
      player2 = player.player2
      expect(player1).not_to eq('George')
      expect(player2).not_to eq('Hommer')
    end`

**Review by george-shammar — CHANGES_REQUESTED** [review]
### **STATUS: CHANGES REQUIRED** :recycle: :recycle: :construction: 

Hello @uzairali19 ,

You're almost there. Please don't be frustrated or irritated. :pray: 

**Highlight**
:heavy_check_mark: You have implemented some of the previous changes. :+1: 

**To improve**

- [ ] As requested by the previous reviewers, you need to implement negative test cases for your test examples. Please see my in-line comment for a typical example of a negative test case for one of your test examples.

**Review by voscarmv — APPROVED** [review]
# Project status: Approved 💯 ✨ 😎 👍 
### You may merge this branch and move forward ⛙🛣️ 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻
# Happy Coding! 🧠 ⌨️ 🖱️ 🖥️

**Pull request merged**